### PR TITLE
Feature hydrogen aluminum integration updates

### PR DESCRIPTION
**Pull request by benson31**
Integrates changes from [this PR](https://github.com/LLNL/Elemental/pull/44) on Hydrogen. This breaks backward compatibility with "pre-#44" hydrogen.

Changes are really minor, but the interface to `lbann::comm` is affected.

If built with Aluminum support in Hydrogen, this should resolve some of the collective issues that various users have been seeing (e.g., `MPI_Reduce_scatter_block`) by just avoiding the MPI calls entirely.

### CODE REVIEW

**Comment by benson31** [comment]
Based on feedback from this meeting, I'm going to add calls that default to dispatching for CPU memory. I didn't do this initially because I was concerned it could trip people up if they work with GPU memory, but in actuality, there aren't many cases in which explicit GPU dispatch is needed inside LBANN.

File: src/layers/transform/in_top_k.cu [hunk 1]
```diff
@@ -240,7 +240,7 @@ void fp_gpu(lbann_comm& comm,
                     top_entries.size() * sizeof(entry),
                     reinterpret_cast<El::byte*>(global_top_entries.data().get()),
                     top_entries.size() * sizeof(entry),
-                    col_comm);
+                    col_comm, El::SyncInfo<El::Device::GPU>{stream, nullptr});
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/layers/transform/in_top_k.cu]: Not a good one. Fixed.

**Comment by benson31** [comment]
I have addressed a few of the comments. The `lbann::comm` interface has been augmented with some routines that assume CPU memory. This could be dangerous if folks are unaware that they should specify. Thoughts?

**Comment by ndryden** [comment]
Document what people should do in terms of CPU vs GPU memory. It would be nice if we could eventually infer it automatically, but that has an overhead. (Maybe we should add a debug mode check that checks whether we pass the right kind of pointers...)

**Comment by benson31** [comment]
Where's the right place to put that documentation? Somewhere in `comm.hpp`, blog post, somewhere else? I don't know where people will have the highest probability of actually seeing it.

**Comment by ndryden** [comment]
I'd put it in `comm.hpp` and mention people should read it in the next blog post. None of these actually have high probability of working, but at least you tried. 🤷‍♂️

**Comment by benson31** [comment]
Re directly passing vs moving `El::mpi::Comm` objects: we should take them by value in the `lbann_comm` interface so it's the user's choice whether or not to increment the reference count on the Aluminum communicator's shared_ptr (recall that Hydrogen stores the Aluminum communicator by shared_ptr to facilitate lazy construction of the Aluminum comm (or no construction for `COMM_NULL`)). When we forward the `El::mpi::Comm` through various layers of the interface, we should move it so the reference count is never incremented once it's inside an `lbann_comm` function stack, since that is an atomic increment/decrement. Likewise when we pass it from the `lbann_comm` stack into the `El::mpi` stack. I'll peruse the interface and clean this up.

**Comment by benson31** [comment]
@ndryden The move stuff in `lbann_comm` should be uniform now.

**Comment by benson31** [comment]
I also put some documentation at the top of `comm.hpp`. Let me know if anything is unclear.

**Comment by ndryden** [comment]
I tested this with ResNet-50 on Pascal and didn't see any issues. It also passed gradient checking in `model_mnist_conv_graph.prototext`.

**Comment by benson31** [comment]
@ndryden Thanks! Anyone else had a chance to test this stuff? It'd be great to get it merged in since there's already a fair body of work building on it.

**Comment by JaeseungYeom** [comment]
Seems ok for Siamese training and transfer learning as well.

**Pull request merged**